### ISSUE TITLE
[BUG] The RestController always consumes content when handling real HTTP requests

### ISSUE THREAD

**Issue by dbwiddis** [opensearch-project/OpenSearch#13011]
TLDR:
 - Behavior was added in https://github.com/opensearch-project/OpenSearch/commit/687978b7d167d4c20a29861b2558b2749c349529 to reject REST requests containing a body, that did not use that body, in January 2019.
 - That behavior was inadvertently disabled for real-world HTTP-sourced requests in https://github.com/opensearch-project/OpenSearch/commit/1cd464d675b3cc3fd478b5b530b2d3a316bc2c6b in August 2019 but remained enabled in testing with mocked clients.
 - Unit tests remain from the first commit, and new tests written for Flow Framework demonstrate that the unit-tested behavior does not match the actual behavior (this is the bug)
 - Solutions involve reverting the first change (since it's effectively disabled, and a poor design IMO) or enabling the first change's intended behavior (which could break a lot of scripts unknowingly relying on the existing behavior). I favor the reverting option but want feedback first.

### Describe the bug

The `BaseRestHandler` includes a check whether content (the REST body) has been consumed by an API, throwing an exception here if it does not do so:

https://github.com/opensearch-project/OpenSearch/blob/c25b00c6b62fd6766999c09fd06dd1f7b281b3eb/server/src/main/java/org/opensearch/rest/BaseRestHandler.java#L125-L127

This exception is tested in OpenSearch [in a few places](https://github.com/search?q=repo%3Aopensearch-project%2FOpenSearch%20%22does%20not%20support%20having%20a%20body%22&type=code).

Unfortunately, it appears that these unit tests which rely on the behavior of the `BaseRestHandler` do not take into account the fact that the `RestController` always consumes content here when a REST request arrives via the HTTP handler:
https://github.com/opensearch-project/OpenSearch/blob/2dc071ff2546b0c6483b9e0a382badf9ac353cdc/server/src/main/java/org/opensearch/rest/RestController.java#L296

It appears that in an actual cluster handling HTTP requests, the above exception is dead code.  

However, it is actively tested directly calling the Rest Handler without using the Rest Controller code path which reads the content.

This is confusing to developers, undocumented and inconsistent behavior, and leads to incorrect assumptions, ultimately returning a confusing 500 error code to end users without any hint of why.

### Related component

Other

### To Reproduce

1. Run OpenSearch with the Flow Framework plugin installed, which relied on this exception handling
2. Query a workflow that does not support a body, but include a body in the request, for example
```json
localhost:9200/_plugins/_flow_framework/workflow/any_id_here
{}
```
3. Get this response:
```json
{
    "error": {
        "root_cause": [
            {
                "type": "illegal_state_exception",
                "reason": "Channel is already closed"
            }
        ],
        "type": "illegal_state_exception",
        "reason": "Channel is already closed"
    },
    "status": 500
}
```

### Expected behavior

Given the multiple tests of the error message, one would suggest the expected behavior is a 400-series REST response with error message: "request [GET /_plugins/_flow_framework/workflow/{workflow_id}] does not support having a body"

However, testing the other OpenSearch APIs with unit tests for this behavior, it appears they silently ignore the body.  In particular, [the commit which added the exception](https://github.com/opensearch-project/OpenSearch/commit/687978b7d167d4c20a29861b2558b2749c349529) indicated that it should have triggered on deleting an index while including a body.

Given that this behavior has existed for 5 years, and the exception was only put in a year before that, and there are probably many hundreds (thousands? millions?) of scripts which might break if we restore the "tested" expected behavior, it seems that perhaps users should "expect" the presence of a body to be silently ignored.

### Additional Details

**Plugins**

Reproduced with only the Flow Framework plugin installed as reported in https://github.com/opensearch-project/flow-framework/issues/638

**Additional context**

I'm happy to submit a PR to "fix" this but would like some input from other maintainers on how they'd like it fixed.  I see two options:

1. In the REST controller content length check, use the overload that does not consume the content, e.g.,
```java
final int contentLength = request.content(false).length();
```
This will restore the "intended" behavior and probably break scripts worldwide.

2. Delete the consumed content check entirely and remove any tests that rely on it; explicitly code this behavior if desired; or appropriately document that content is ignored.

**Comment by cwperks** [comment]
Why is the channel already closed? Are there multiple places trying to send a response on the RestChannel? Once a response is sent on the RestChannel (like [this](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/filter/SecurityRestFilter.java#L135)) the channel is closed and any attempt to send another response will receive a `Channel is already closed` error.

Does the check make sense where it is in the first place? I don't believe many actions will read the request content inside of prepareRequest, but instead will read them when the action is accepted [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/rest/BaseRestHandler.java#L131).

**Comment by dbwiddis** [comment]
> Why is the channel already closed? 

Null pointer exception in our (my) implementation. I returned null never expecting it to be read because it would have hit that other illegal argument exception. 

> Does the check make sense where it is in the first place? 

No, I never liked it because it allowed you to go through all of the actions you would normally take, and only gave an error at the end after the response, swallowing/hiding any actual response.  I disliked where it was so much I coded up a fail-fast (but didn't realize it didn't actually have any effect other than in tests).

> I don't believe many actions will read the request content inside of prepareRequest, but instead will read them when the action is accepted [here](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/rest/BaseRestHandler.java#L131).

It varies. The content sometimes needs to be parsed on the Rest Action side before being sent over transport.  Many plugins do this.

**Comment by dbwiddis** [comment]
To be clear, the bug is that there's a block of dead code (in actual operation) that's only triggered on unit tests that is confusing to devs like me. :)

**Comment by peternied** [comment]
[Triage - attendees [1](https://github.com/lukas-vlcek) [2](https://github.com/cwperks) [3](https://github.com/jed326) [4](https://github.com/scrawfor99) [5](https://github.com/kotwanikunal) [6](https://github.com/srikanthpadakanti) [7](https://github.com/peternied) [8](https://github.com/andrross)]
@dbwiddis Thanks for filing
